core/primitives/src/epoch_manager.rs

```diff
@@ -615,7 +655,7 @@ pub mod epoch_info {
                 let block_producers_sampler = stake_weights(&block_producers_settlement);
                 let chunk_producers_sampler =
                     chunk_producers_settlement.iter().map(|vs| stake_weights(vs)).collect();
-                Self::V3(EpochInfoV3 {
+                Self::V4(EpochInfoV4 {
```

Comment:
The other changes are fine, but this one cannot just be changed from V3 to V4. We would have to introduce a new (unstable) protocol feature and then check the protocol feature to decide whether to return V3 or V4. You can look at protocol_upgrade.md to learn about that procedure. It would involve introducing a nightly protocol feature and then guarding the V4 creation here. You don't need to worry about launching the feature - we just need to gate it for now. Let me know if you have questions here.

Reply:
I had changed it to verify that `EpochInfoV4` can be wired in without causing issues. Though now I understand that this change must go through a separate process. Thanks for the and explanation and reference!

In bc7bc1d I reverted `EpochInfo::new` to return `V3` by default. Should the feature gate be introduced as part of this PR or later on in a separate PR?

Reply:
I'm OK with doing the feature gate in a separate PR. I had mentioned it because without the feature gate, I don't feel that the test you introduced in validator_selection.rs (which is a very good test) would pass, as you would be constructing a EpochInfoV3. You would only get a EpochInfoV4 if you passed in a protocol version where the feature is enabled.

Reply:
Sorry for the failing test. It slipped through locally and somehow after pushing updates an approval was required for tests to be run in CI (IIRC they ran right away when the PR was opened). The test failure is indeed due to `EpochInfo::new` not returning `V4` anymore. So I’ll work on adding the feature gate with this PR and then include the test only when `V4` is returned.

Reply:
Added the feature-gate to return `EpochInfo::V4` when the feature `protocol_feature_chunk_validation` is enabled in 8378c80. The corresponding enum variant is `ProtocolFeature::ChunkValidation`.